ADR/0033-product-version-controller.md

```diff
@@ -0,0 +1,493 @@
+example, access credentials for a Git repository need to be filled-in every time
+a component referring to that repository is created, even if other components
+referring that repository already exist in the user's workspace.
+
+The intention for the solution proposed in this document is to follow the
+pattern described above as far as the Konflux "Application" and "Component"
+objects go and provide a higher-level layer for the streamlined management of
+large amounts of such objects.
+
+An alternative approach to managing the challenges we have described above is to
+modify the way Application and Component object function so that for example:
+
+* A component may encompass multiple versions of similar code across different
+  Git branches
+* An application may include several sub groups of components so that each could
+  be regarded to be defining a different version
+
+We see several benefits to the "high level layer" approach we propose over the
+alternative:
+
+* The way Components and Applications work is currently tightly coupled to the
+  way the various Konflux services work. Changing it would require changes
+  across multiple services and may introduce operational risks.
+* There are proposals being discussed for decoupling the various Konflux
+  services and possibly deprecating the Component and Application objects.
+  Having a separate high-level layer would allow for adjusting it for using
+  different underlying objects while continued emphasis on using Application
+  and Component object would necessitate continued maintenance of them.
+
+## Decision
+
+### Facilitating product and version based navigation in the UI
+
+We will define two new custom resources:
+
+**Project** resources will define a large project that can have multiple
+parallel development streams that may be each released as a different product
+version.
+
+**ProjectDevelopmentStream** resources will define a stream of development
+within a *Project*. It will be linked to an owning project via a [Kubernetes
+owner reference][oref].
+
+*Application*, *Component*, and any other Konflux object that needs to be
+created to facilitate the development of a particular project development stream
+will be linked to a *ProjectDevelopmentStream* object via an owner reference.
+
+**Note:** The choice was made to use the terms "Project" and "Project
+Development Stream" as opposed to "Product" and "Version" because those can be
+applied more broadly also to upstream community projects as opposed to
+marketable products. Those terms also more accurately describe the environment
+in which software development happens. [See here][pvsp] for a more detailed
+discussion of these terms.
+
+[oref]: https://kubernetes.io/docs/concepts/overview/working-with-objects/owners-dependents/
+[pvsp]: https://opensource.com/article/19/11/product-vs-project
+
+### Streamlining version creation from the API/CLI.
+
+We will define the **ProjectDevelopmentStreamTemplate** custom resource. The
```

Comment:
Assume we do want and need `Project` and `ProjectDevelopmentStream` for a moment.

Do we need `ProjectDevelopmentStreamTemplate`?

* Is it possible to re-use openshift templates rather than add another API resource here? We need our system to be able to work on kube, not just on openshift - so that may nix that option.
* Is it possible to use helm with in-cluster resources like it is with openshift templates?

Reply:
While I am looking at OpenShift templates to get ideas about how to implement this. I think the scope for ProjectDevelopmentStreamTemplates is much smaller then for Helm of OpenShift templates. I think the set of capabilities needs to be much narrower.

Reply:
Reflecting conclusions from conversations that happened elsewhere - there is no readily-available solution that is commonly used across the K8s landscape and that would fit our purpose, so we'll have to make our own.